### PR TITLE
Combobox for form entry

### PR DESCRIPTION
**Pull request by benrudolph**
@wpride had a little fun implementing the combobox for web entry :sweat_smile: i had initially tried to do this with atwho.js, but it just isn't the right tool for what we're trying to do. the jquery typeahead library has some great capabilities. you can read the docs on combobox/behavior here: https://confluence.dimagi.com/display/commcarepublic/Advanced+CommCare+Android+Formatting#AdvancedCommCareAndroidFormatting-SingleSelect"ComboBox"

screen shot 2017-01-27 at 9 49 08 am
screen shot 2017-01-27 at 9 48 52 am
screen shot 2017-01-27 at 9 49 03 am

cc: @emord @snopoke

design: @orangejenny

### CODE REVIEW

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 1]
```diff
@@ -321,6 +321,78 @@ DropdownEntry.prototype.onPreProcess = function(newValue) {
     }
 };
 
+function ComboboxEntry(question, options) {
+    var self = this;
+    EntrySingleAnswer.call(this, question, options);
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: 'EntrySingleAnswer' is not defined. (no-undef)

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 2]
```diff
@@ -321,6 +321,78 @@ DropdownEntry.prototype.onPreProcess = function(newValue) {
...
+    } else if (matchType === Formplayer.Const.COMBOBOX_FUZZY) {
+        // Fuzzy filter, matches if any of the query is in choice
+        return d.display.includes(query);
+    } else {
+        // Standard filter, matches only start of word
+        return d.display.startsWith(query);
+    }
+};
+
+ComboboxEntry.prototype = Object.create(EntrySingleAnswer.prototype);
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: 'EntrySingleAnswer' is not defined. (no-undef)

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 3]
```diff
@@ -321,6 +321,78 @@ DropdownEntry.prototype.onPreProcess = function(newValue) {
...
+ComboboxEntry.prototype.constructor = EntrySingleAnswer;
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: 'EntrySingleAnswer' is not defined. (no-undef)

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 4]
```diff
@@ -321,6 +321,86 @@ DropdownEntry.prototype.onPreProcess = function(newValue) {
...
+/**
+ * The ComboboxEntry is an advanced android formatting entry. It is enabled
+ * when the user specifies combobox in the appearance attributes for a
+ * single select question.
+ *
+ * Docs: https://confluence.dimagi.com/display/commcarepublic/Advanced+CommCare+Android+Formatting#AdvancedCommCareAndroidFormatting-SingleSelect"ComboBox"
+ */
+function ComboboxEntry(question, options) {
+    var self = this;
+    EntrySingleAnswer.call(this, question, options);
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: 'EntrySingleAnswer' is not defined. (no-undef)

**Comment by wpride** [comment]
nice! cc @amstone326

**Review by wpride — APPROVED** [review]

**Comment by wpride** [comment]
@benrudolph test failure looks real

**Comment by orangejenny** [comment]
Design looks fine to me, though I'm curious what was needed that's in jquery typeahead but not in atwho/select2.

**Comment by benrudolph** [comment]
@orangejenny atwhojs is really only good for progressively adding autocomplete fields. like slack emojs. it's hard to restrict it to _just_ one slack emoji and nothing else, which is what i'm attempting to do here. and the space that it adds after you enter an autocomplete field is killer

**Comment by orangejenny** [comment]
Vellum uses atwho as its "autocomplete" (it allowed us to get rid of jquery ui without having to introduce select2 or yet another library) and handles both of those issues. To play with it, add a lookup table and click on the `Lookup Table Data` node; the `Value Field` and `Display Text Field` use it in this manner.

The config: https://github.com/dimagi/Vellum/blob/master/src/atwho.js#L54-L74

**Comment by amstone326** [comment]
Hey @benrudolph @wpride I apologize for not making this clearer initially, but I had intended for you to be able to reuse the `ComoboxFilterRule` classes that are in core when implementing this on formplayer, so that the filter behaviors between this and mobile would be automatically aligned without having to reimplement any filter code. Does it seem like a big lift to integrate that in here? There's 1 implementation of a filter rule for each type of combobox, and the class exposes a `choiceShouldBeShown(String choice, CharSequence textEntered)` method that returns whether or not a specific choice should be included in the visible answer options based on the text entered by the user.

**Comment by benrudolph** [comment]
@orangejenny interesting... that does seem to do what i want. i can try and switch it back to that, thanks (though depending on aliza's comment it may not make sense)

@amstone326 we can do that, however it would require an ajax request every time someone types a character. in this version we don't make a request to the server until they've entered a valid choice. i'm kind of torn as to whether being DRY here is better since it requires a fairly big lift given the filter logic isn't crazy complicated

**Comment by amstone326** [comment]
@benrudolph Ah ok, did not realize that. It sounds like that's not the best idea then, but as a compromise I think it would be worthwhile to try to make the separate implementations equivalent, which I think they aren't quite right now. The "standard" one looks right to me, but for the others:

-The "fuzzy" implementation on mobile uses a [util method we have](https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/commcare/cases/util/StringUtils.java#L69-L137) for fuzzy-matching strings, which means that options that are close-but-not-quite matches will be shown. Would it make sense to just straight up copy over the logic of that code to formplayer and use it here? I'm particularly worried about this one having parity because this is the version that UATBC has elected to use, and they were specifically concerned about having the fuzzy-matching behavior because they said they will be using the combo box with answer choices that are commonly misspelled (complex location names).

-For the "multiword" implementation, I could be misreading the code but I think what you implemented is sort of the inverse of what I have. On mobile, what it does is split the _query_ up into words and then shows any answer choice which contains all of those words. Whereas it looks like you're splitting the _answer choice_ up into words and then checking if the query _starts with_ any of those words. Let me know if I'm misreading that one though.

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 5]
```diff
@@ -321,6 +321,86 @@ DropdownEntry.prototype.onPreProcess = function(newValue) {
...
+    } else if (matchType === Formplayer.Const.COMBOBOX_FUZZY) {
+        // Fuzzy filter, matches if any of the query is in choice
+        return d.display.includes(query);
+    } else {
+        // Standard filter, matches only start of word
+        return d.display.startsWith(query);
+    }
+};
+
+ComboboxEntry.prototype = Object.create(EntrySingleAnswer.prototype);
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: 'EntrySingleAnswer' is not defined. (no-undef)

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 6]
```diff
@@ -321,6 +321,86 @@ DropdownEntry.prototype.onPreProcess = function(newValue) {
...
+ComboboxEntry.prototype.constructor = EntrySingleAnswer;
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: 'EntrySingleAnswer' is not defined. (no-undef)

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 7]
```diff
@@ -321,6 +321,86 @@ DropdownEntry.prototype.onPreProcess = function(newValue) {
...
+        this.answer(value.id);
+        this.question.error(null);
+    } else {
+        this.question.error(gettext('Not a valid choice'));
+    }
+};
+
 $.datetimepicker.setDateFormatter({
     parseDate: function (date, format) {
         var d = moment(date, format);
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: 'moment' is not defined. (no-undef)

**Comment by benrudolph** [comment]
@orangejenny -- https://github.com/dimagi/commcare-hq/pull/14673/commits/8eeed4cc10afcad6ddde00af29706cbcb1b1cf98 switched back to atwho. for some reason i constantly miss new settings on the docs page even though i think i've read them all. thanks for pointing me in the right direction

@amstone326 -- https://github.com/dimagi/commcare-hq/pull/14673/commits/099ef9e4812bbecdbd748ff217aad0763b33419d ok took another go at the filters. i think i'm doing what you've said now. i decided not to always return `false` if the query was less than 3 for the fuzzy match though because i thought that would then mean you could never match 3 letter choices? unless you guys have more logic that matches when it's an exact match

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 8]
```diff
@@ -321,6 +321,112 @@ DropdownEntry.prototype.onPreProcess = function(newValue) {
...
+    self.renderAtwho = function() {
+        var $input = $('#' + self.entryId);
+        $input.atwho({
+            at: '',
+            data: self.options(),
+            maxLen: Infinity,
+            tabSelectsMatch: false,
+            suffix: '',
+            callbacks: {
+                filter: function(query, data, searchKey) {
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: 'searchKey' is defined but never used (no-unused-vars)

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 9]
```diff
@@ -321,6 +321,112 @@ DropdownEntry.prototype.onPreProcess = function(newValue) {
...
+                    return _.filter(data, function(item) {
+                        return ComboboxEntry.filter(query, item, self.matchType);
+                    });
+                },
+                matcher: function(flag, subtext, should_startWithSpace) {
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: 'should_startWithSpace' is defined but never used (no-unused-vars)

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 10]
```diff
@@ -321,6 +321,112 @@ DropdownEntry.prototype.onPreProcess = function(newValue) {
...
+                },
+            },
+        });
+    };
+    self.isValid = function(value) {
+        if (!value) {
+            return true;
+        }
+        return _.include(
+            _.map(self.options(), function(option) { return option.name }),
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: Missing semicolon. (semi)

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 11]
```diff
@@ -321,6 +321,112 @@ DropdownEntry.prototype.onPreProcess = function(newValue) {
...
+    } else if (matchType === Formplayer.Const.COMBOBOX_FUZZY) {
+        // Fuzzy filter, matches if query is "close" to answer
+        return window.Levenshtein.get(d.name.toLowerCase(), query.toLowerCase()) <= 2;
+    } else {
+        // Standard filter, matches only start of word
+        return d.name.startsWith(query);
+    }
+};
+
+ComboboxEntry.prototype = Object.create(EntrySingleAnswer.prototype);
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: 'EntrySingleAnswer' is not defined. (no-undef)

**Comment by orangejenny** [comment]
Sweet, thank you for reworking it with atwho, admit I tend to be draconian about adding js libraries.

**Comment by amstone326** [comment]
@benrudolph Re: your last comment, yea I'm using this https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/javarosa/core/model/FuzzyMatchFilterRule.java#L56-L60 to check for either a fuzzy match OR exact substring, so I think it makes sense to implement Levenshtein the standard way and then check for either as well. (Side note: it looks like you're just importing a pre-existing javascript implementation of Levenshtein in this PR, so how did you change the behavior anyway? My impression was that Levenshtein distance is a standard formula).

Other than that, looks good to me 👍 Thanks for changing

**Comment by benrudolph** [comment]
@amstone326 cool, that's easy to add.

> Side note: it looks like you're just importing a pre-existing javascript implementation of Levenshtein in this PR, so how did you change the behavior anyway? My impression was that Levenshtein distance is a standard formula

I don't understand this part. I am using a pre existing library since it's just a standard algo as you mentioned. I didn't change the functionality of it at all

**Comment by amstone326** [comment]
@benrudolph Sorry, realized I misunderstood your last comment, can ignore that question.

**Comment by benrudolph** [comment]
good to merge

**Pull request merged**